Alias delete

AsyaAndKo opened the pull request:
Create deleteUrlAlias function for UrlRepositoryFakeImpl, provide tests and @Delete endpoint for AuthenticatedApiController.

AsyaAndKo committed:
Add deleteUrlAlias function and tests

Add delete endpoint but no tests because micronaut tests are scary.
Maybe next time (no)?

AsyaAndKo committed:
Fix formatting

AsyaAndKo committed:
Fix some more formatting errors

AsyaAndKo committed:
Starting point to hate java formatting

AsyaAndKo committed:
Add deleteAlias in Logic class

OlehKyba commented on src/main/java/edu/kpi/testcourse/rest/AuthenticatedApiController.java:
```diff
@@ -30,23 +31,27 @@
-      HttpHostResolver httpHostResolver
-  ) {
+      HttpHostResolver httpHostResolver,
+      UrlRepository urlRepository) {
```
In my opinion, you should not use the Url Repository instance directly in the controller. You should create a separate method for the deletion in the Logic class.

OlehKyba commented on src/main/java/edu/kpi/testcourse/rest/AuthenticatedApiController.java:
```diff
@@ -71,4 +76,23 @@ public AuthenticatedApiController(
+                                Principal principal) throws IllegalArgumentException {
+    try {
+      urlRepository.deleteUrlAlias(request.alias(), principal.getName());
+      return HttpResponse.ok();
```
According to the specification, after success deletion, we must return 204 HTTP code. So you should change this row on:
```java
return HttpResponse.noContent();
```

OlehKyba replied:
```suggestion
      return HttpResponse.noContent();
```

OlehKyba commented on src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java:
```diff
@@ -25,10 +28,33 @@ public void createUrlAlias(UrlAlias urlAlias) {
+   * @param email of the user to whom the UrlAlias belongs
+   * @param alias the UrlAlias which should be removed
+   *
+   * @throws IllegalArgumentException if no such email or alias
```
According to the [repository interface](https://github.com/future-stardust/url-shrtnr-palevo/blob/main/src/main/java/edu/kpi/testcourse/storage/UrlRepository.java): 

1. RuntimeException if the repository does not contain a URL alias with this short name;
2. PermissionDenied if the UrlAlias doesn't belong to the current user.

OlehKyba commented on src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java:
```diff
@@ -25,10 +28,33 @@ public void createUrlAlias(UrlAlias urlAlias) {
   public void deleteUrlAlias(String email, String alias) {
-    // TODO: We should implement it
-    throw new UnsupportedOperationException();
+    UrlAlias foundUrlAlias = null;
```
You use the hash map, so you don't need the for statement to find element by alias.
```suggestion
    UrlAlias foundUrlAlias = aliases.get(alias);
```

OlehKyba commented on src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java:
```diff
@@ -25,10 +28,33 @@ public void createUrlAlias(UrlAlias urlAlias) {
+      }
+    }
+    if (foundUrlAlias == null) {
+      throw new IllegalArgumentException();
```
RuntimeException if the repository does not contain a URL alias with this short name.
```suggestion
      throw new RuntimeException();
```

OlehKyba commented on src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java:
```diff
@@ -25,10 +28,33 @@ public void createUrlAlias(UrlAlias urlAlias) {
+    if (foundUrlAlias == null) {
+      throw new IllegalArgumentException();
+    }
+    if (aliases.containsKey(alias)) {
```
If we can find UrlAlias by the alias, the UrlAlias exists in the hash map. Instead, you need to check if this alias belongs to the user who is trying to delete it.
```suggestion
   if (!alias.email().equals(email)) {
      throw new PermissionDenied();
    }
```

AsyaAndKo replied:
I didn't quite understand what do you mean by alias.email
Maybe like this then: !aliases.get(email).equals(email)

AsyaAndKo replied:
I changed the function, now it throws all needed exceptions, check it out

OlehKyba replied:
:+1:

OlehKyba commented on src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java:
```diff
@@ -2,14 +2,17 @@
-  private final HashMap<String, UrlAlias> aliases = new HashMap<>();
+
+  //Make it public to make my testing process less painful
+  public final HashMap<String, UrlAlias> aliases = new HashMap<>();
```
You cannot do this! You should use only public API in the tests. Instead of a hash map field, use in tests methods like findUrlAlias.

OlehKyba commented on src/test/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImplTest.java:
```diff
@@ -38,5 +38,38 @@ void shouldNotAllowToCreateSameAliases() {
+    repo.deleteUrlAlias(alias1.email(), alias1.alias());
+
+    //THEN
+    assertFalse(repo.aliases.containsValue(alias1));
```
```suggestion
    assertThat(repo.findUrlAlias("http://r.com/short")).isNull();
```

AsyaAndKo committed:
Add deleteUrlAlias to Logic class

Now it's possible to call deleteUrlAlias from logic class in
AuthenticatedApiContoller. Also change
HttpResponse.ok() -> HttpResponse.noContent()

AsyaAndKo committed:
Change deleteUrlAlias function to be able to throw
 RuntimeException and PermissionDenied errors.

Use findUrlAlias function in deleteUrlAlias function.
Not sure if it's a good idea but let it be.
Add tests for RuntimeExeption and PermissionDenied errors.

AsyaAndKo committed:
Some checkstyle fixes

YustinaOvcharova commented on src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java:
```diff
@@ -25,10 +27,30 @@ public void createUrlAlias(UrlAlias urlAlias) {
-    throw new UnsupportedOperationException();
+    UrlAlias foundUrlAlias = findUrlAlias(alias);
+
+    if (foundUrlAlias == null || foundUrlAlias.email() == null) {
```
Redundant if statement part (foundUrlAlias.email() == null)
```suggestion
    if (foundUrlAlias == null) {
         throw new RuntimeException();
    }
```

AsyaAndKo committed:
Update src/main/java/edu/kpi/testcourse/storage/UrlRepositoryFakeImpl.java

Delete redundant if statement part

Co-authored-by: Yustina Ovcharova <45105945+YustinaOvcharova@users.noreply.github.com>

AsyaAndKo committed:
Update UrlRepositoryFakeImpl.java

Delete redundant if statement part

OlehKyba reviewed (approved):
LGTM :+1:
